src/passes/TypeSSA.cpp

```diff
@@ -147,14 +148,136 @@ std::vector<HeapType> ensureTypesAreInNewRecGroup(RecGroup recGroup,
 // A vector of struct.new or one of the variations on array.new.
 using News = std::vector<Expression*>;
 
-struct NewFinder : public PostWalker<NewFinder> {
+// A set of types for which the exactness of allocations may be observed.
+using ExactTypes = std::unordered_set<HeapType>;
+
+struct Analyzer
+  : public PostWalker<Analyzer, UnifiedExpressionVisitor<Analyzer>> {
+  // Find allocations we can potentially optimize.
   News news;
 
-  void visitStructNew(StructNew* curr) { news.push_back(curr); }
-  void visitArrayNew(ArrayNew* curr) { news.push_back(curr); }
-  void visitArrayNewData(ArrayNewData* curr) { news.push_back(curr); }
-  void visitArrayNewElem(ArrayNewElem* curr) { news.push_back(curr); }
-  void visitArrayNewFixed(ArrayNewFixed* curr) { news.push_back(curr); }
+  // Also find heap types for which the exactness of allocations is observed. We
+  // will not be able to optimize allocations of these types without an analysis
+  // proving that an allocation does not flow into any location where its
+  // exactness is observed.
+  ExactTypes exactTypes;
```

Comment:
Maybe "disallowedExactTypes", or "disallowedTypes", as this is not the list of all exact types?